pydantic_ai_slim/pydantic_ai/ui/vercel_ai/_adapter.py

```diff
@@ -197,3 +231,204 @@ def load_messages(cls, messages: Sequence[UIMessage]) -> list[ModelMessage]:  #
+                            thinking_metadata['signature'] = part.signature
+                        if part.provider_name is not None:
+                            thinking_metadata['provider_name'] = part.provider_name
+                        if part.provider_details is not None:
+                            thinking_metadata['provider_details'] = part.provider_details
+
+                        provider_metadata = {'pydantic_ai': thinking_metadata} if thinking_metadata else None
+                        ui_parts.append(
+                            ReasoningUIPart(text=part.content, state='done', provider_metadata=provider_metadata)
+                        )
+                    elif isinstance(part, FilePart):
+                        ui_parts.append(
+                            FileUIPart(
+                                url=part.content.data_uri,
+                                media_type=part.content.media_type,
+                            )
+                        )
+                    elif isinstance(part, BaseToolCallPart):
+                        if isinstance(part, BuiltinToolCallPart):
+                            prefixed_id = (
+                                f'{BUILTIN_TOOL_CALL_ID_PREFIX}|{part.provider_name or ""}|{part.tool_call_id}'
+                            )
+
+                            if builtin_return := local_builtin_returns.get(part.tool_call_id):
+                                content = builtin_return.model_response_str()
+                                call_provider_metadata = (
+                                    {'pydantic_ai': {'provider_name': part.provider_name}}
+                                    if part.provider_name
+                                    else None
+                                )
+                                ui_parts.append(
+                                    ToolOutputAvailablePart(
+                                        type=f'tool-{part.tool_name}',
+                                        tool_call_id=prefixed_id,
+                                        input=part.args_as_json_str(),
+                                        output=content,
+                                        state='output-available',
+                                        provider_executed=True,
+                                        call_provider_metadata=call_provider_metadata,
+                                    )
+                                )
+                            else:  # pragma: no cover
+                                ui_parts.append(
+                                    ToolInputAvailablePart(
+                                        type=f'tool-{part.tool_name}',
+                                        tool_call_id=prefixed_id,
+                                        input=part.args_as_json_str(),
+                                        state='input-available',
+                                        provider_executed=True,
```

Comment:
Shouldn't we still store the metadata in this case?

Reply:
yeahp we should, I'm not even sure this case is reachable but I should've included the metadata